File: mybus.app/scripts/update_gtfs.R
## update the GTFS stuff
library(RPostgreSQL)
library(jsonlite)
library(caroline)

drv = dbDriver("PostgreSQL")
con = dbConnect(drv, dbname = "homestead", host = "localhost", user = "homestead", port = "54320", password = "secret")

system('curl -s -X GET "https://api.at.govt.nz/v2/gtfs/versions" -H "Ocp-Apim-Subscription-Key: 1da1fb03bbed453ca7e429643e6abad1" -o gtfs_versions.json')
versions = fromJSON('gtfs_versions.json')$response
unlink('gtfs_versions.json')

curVer = dbGetQuery(con, "SELECT version FROM gtfs_versions")
new = which(! versions$version %in% curVer$version)
if (length(new) > 0) {
  dbWriteTable(con, "gtfs_versions", versions[new, ], append = TRUE, row.names = FALSE)


  ## Now just go through each file and update things:

  ## --------------------------------------------------------------- AGENCIES
  system('curl -s -X GET "https://cdn01.at.govt.nz/data/agency.txt" -o agencies.csv')
  agencies = read.csv("agencies.csv")
  unlink("agencies.csv")
  curAg = dbGetQuery(con, "SELECT agency_id FROM agencies")
  new = which(! agencies$agency_id %in% curAg$agency_id)
  if (length(new) > 0) {
    agencies = agencies[new, ]
    a = data.frame(
      agency_id = agencies$agency_id,
      name = agencies$agency_name,
      url = agencies$agency_url,
      timezone = agencies$agency_timezone,
      lang = agencies$agency_lang,
      phone = agencies$agency_phone
    )

    dbGetQuery(con,
      sprintf("INSERT INTO agencies (agency_id, name, url, timezone, lang, phone) VALUES ('%s')",
              paste(a$agency_id, a$name, a$url, a$timezone, a$lang, a$phone, sep = "','", collapse = "'), ('")))
  }

  ## --------------------------------------------------------------- ROUTES
  system('curl -s -X GET "https://cdn01.at.govt.nz/data/routes.txt" -o routes.csv')
  routes = read.csv("routes.csv")
  unlink("routes.csv")
  curAg = dbGetQuery(con, "SELECT id FROM routes")
  new = which((! routes$route_id %in% curAg$id) &
              grepl(paste0("-", versions$version, collapse="|"), routes$route_id))
  if (length(new) > 0) {
    routes = routes[new, ]
    a = list(
      id = routes$route_id,
      route_id = gsub(paste0("-", versions$version, collapse="|"), "", routes$route_id),
      short_name = routes$route_short_name,
      long_name = routes$route_long_name,
      type_id = routes$route_type
    )
    a$version = sapply(1:nrow(routes), function(i) gsub(paste0(a$route_id[i], "-"), "", a$id[i]))
    if (!is.null(routes$agency_id)) a$agency_id =  routes$agency_id
    if (!is.null(routes$route_desc)) a$desc = routes$route_desc
    if (!is.null(routes$route_url)) a$url = routes$route_url
    if (!is.null(routes$route_color)) a$color = routes$route_color
    if (!is.null(routes$route_text_color)) a$text_color = routes$route_text_color

    dbGetQuery(con,
      sprintf("INSERT INTO routes (%s) VALUES ('%s')",
              paste(names(a), collapse = ","),
              do.call(paste, c(a, list(sep = "','", collapse = "'),('")))))
  }

  ## --------------------------------------------------------------- CALENDARS
  system('curl -s -X GET "https://cdn01.at.govt.nz/data/calendar.txt" -o calendar.csv')
  cal = read.csv("calendar.csv")
  unlink("calendar.csv")
  curAg = dbGetQuery(con, "SELECT id FROM calendars")
  new = which((! cal$service_id %in% curAg$id) &
              grepl(paste0("-", versions$version, collapse="|"), cal$service_id))
  if (length(new) > 0) {
    cal = cal[new, ]
    a = list(
      id = cal$service_id,
      service_id = gsub(paste0("-", versions$version, collapse="|"), "", cal$service_id),
      monday = cal$monday == 1,
      tuesday = cal$tuesday == 1,
      wednesday = cal$wednesday == 1,
      thursday = cal$thursday == 1,
      friday = cal$friday == 1,
      saturday = cal$saturday == 1,
      sunday = cal$sunday == 1,
      start_date = cal$start_date,
      end_date = cal$end_date
    )
    a$version = sapply(1:nrow(cal), function(i) gsub(paste0(a$service_id[i], "-"), "", a$id[i]))

    dbGetQuery(con,
      sprintf("INSERT INTO calendars (%s) VALUES ('%s')",
              paste(names(a), collapse = ","),
              do.call(paste, c(a, list(sep = "','", collapse = "'),('")))))
  }

  ## --------------------------------------------------------------- CALENDAR DATES
  system('curl -s -X GET "https://cdn01.at.govt.nz/data/calendar_dates.txt" -o calendar_dates.csv')
  cal = read.csv("calendar_dates.csv")
  unlink("calendar_dates.csv")
  curAg = dbGetQuery(con, "SELECT service_id, date FROM calendar_dates")
  new = which((! paste(cal$service_id, cal$date) %in% paste(curAg$service_id, curAg$date)) &
              grepl(paste0("-", versions$version, collapse="|"), cal$service_id))
  if (length(new) > 0) {
    cal = cal[new, ]
    a = list(
      service_id = cal$service_id,
      date = cal$date,
      exception_type = cal$exception_type
    )

    dbGetQuery(con,
      sprintf("INSERT INTO calendar_dates (%s) VALUES ('%s')",
              paste(names(a), collapse = ","),
              do.call(paste, c(a, list(sep = "','", collapse = "'),('")))))
  }


  ## --------------------------------------------------------------- TRIPS
  system('curl -s -X GET "https://cdn01.at.govt.nz/data/trips.txt" -o trips.csv')
  trips = read.csv("trips.csv")
  unlink("trips.csv")
  curAg = dbGetQuery(con, "SELECT id FROM trips")
  new = which((! trips$trip_id %in% curAg$id) &
              grepl(paste0("-", versions$version, collapse="|"), trips$trip_id))
  if (length(new) > 0) {
    trips = trips[new, ]
    a = list(
      id = trips$trip_id,
      route_id = trips$route_id,
      service_id = trips$service_id,
      trip_id = gsub(paste0("-", versions$version, collapse="|"), "", trips$trip_id)
    )
    a$version = sapply(1:nrow(trips), function(i) gsub(paste0(a$trip_id[i], "-"), "", a$id[i]))
    if (!is.null(trips$trip_headsign)) a$headsign = trips$trip_headsign
    if (!is.null(trips$trip_short_name)) a$short_name = trips$trip_short_name
    if (!is.null(trips$trip_long_name)) a$long_name = trips$trip_long_name
    if (!is.null(trips$direction_id)) a$direction_id = trips$direction_id
    if (!is.null(trips$block_id)) a$block_id = trips$block_id
    if (!is.null(trips$shape_id)) a$shape_id = trips$shape_id
    if (!is.null(trips$wheelchair_accessible)) a$wheelchair_accessible = trips$wheelchair_accessible
    if (!is.null(trips$bikes_allowed)) a$bikes_allowed = trips$bikes_allowed

    dbGetQuery(con,
      sprintf("INSERT INTO trips (%s) VALUES ('%s')",
              paste(names(a), collapse = ","),
              do.call(paste, c(a, list(sep = "','", collapse = "'),('")))))
  }



  ## --------------------------------------------------------------- SHAPES
  system('curl -s -X GET "https://cdn01.at.govt.nz/data/shapes.txt" -o shapes.csv')
  shapes = read.csv("shapes.csv")
  unlink("shapes.csv")
  curAg = dbGetQuery(con, "SELECT id FROM shapes")
  new = which((! shapes$shape_id %in% curAg$id) &
              grepl(paste0("-", versions$version, collapse="|"), shapes$shape_id))
  if (length(new) > 0) {
    shapes = shapes[new, ]
    a = list(
      id = shapes$shape_id,
      shape_id = gsub(paste0("-", versions$version, collapse="|"), "", shapes$shape_id),
      lat = shapes$shape_pt_lat,
      lon = shapes$shape_pt_lon,
      pt_sequence = shapes$shape_pt_sequence
    )
    a$version = sapply(1:nrow(shapes), function(i) gsub(paste0(a$shape_id[i], "-"), "", a$id[i]))
    # if (!is.null(shapes$shape_dist_traveled)) a$dist_traveled = shapes$shape_dist_traveled

    shapeids = unique(shapes$shape_id)
    pb <- txtProgressBar(0, length(shapeids), 0, style = 3)
    for (i in 1:length(shapeids)) {
      setTxtProgressBar(pb, i)
      aa = lapply(a, function(x) x[shapes$shape_id == shapeids[i]])
      dbGetQuery(con,
        sprintf("INSERT INTO shapes (%s) VALUES ('%s')",
                paste(names(a), collapse = ","),
                do.call(paste, c(aa, list(sep = "','", collapse = "'),('")))))
    }
    close(pb)
  }

  ## --------------------------------------------------------------- STOPS
  system('curl -s -X GET "https://cdn01.at.govt.nz/data/stops.txt" -o stop.csv')
  stops = read.csv("stop.csv")
  unlink("stop.csv")
  curAg = dbGetQuery(con, "SELECT id FROM stops")
  new = which((! stops$stop_id %in% curAg$id))
  if (length(new) > 0) {
    stops = stops[new, ]
    a = list(
      id = stops$stop_id,
      name = gsub("'", "''", stops$stop_name),
      lat = stops$stop_lat,
      lon = stops$stop_lon
    )
    # if (!is.null(stops$stop_code)) a$code = stops$stop_code
    # if (!is.null(stops$stop_desc)) a$desc = stops$stop_desc
    # if (!is.null(stops$zone_id)) a$zone_id = stops$zone_id
    # if (!is.null(stops$stop_url)) a$url = stops$stop_url
    # if (!is.null(stops$location_type)) a$location_type = stops$location_type
    # if (!is.null(stops$parent_station)) a$parent_station = stops$parent_station
    # if (!is.null(stops$timezone)) a$timezone = stops$timezone
    # if (!is.null(stops$wheelchair_boarding)) a$wheelchair_boarding = stops$wheelchair_boarding

    dbGetQuery(con,
      sprintf("INSERT INTO stops (%s) VALUES ('%s')",
              paste(names(a), collapse = ","),
              do.call(paste, c(a, list(sep = "','", collapse = "'),('")))))
  }


  ## --------------------------------------------------------------- STOP TIMES
  system('curl -s -X GET "https://cdn01.at.govt.nz/data/stop_times.txt" -o stop.csv')
  stops = read.csv("stop.csv")
  unlink("stop.csv")
  curAg = dbGetQuery(con, "SELECT trip_id FROM stop_times")
  new = which((! stop$trip_id %in% curAg$trip_id) &
              grepl(paste0("-", versions$version, collapse="|"), stops$trip_id))
  if (length(new) > 0) {
    stops = stops[new, ]
    a = list(
      trip_id = stops$trip_id,
      arrival_time = stops$arrival_time,
      departure_time = stops$departure_time,
      stop_id = stops$stop_id,
      stop_sequence = stops$stop_sequence
    )
    # if (!is.null(stops$stop_headsign)) a$stop_headsign = stops$stop_headsign
    # if (!is.null(stops$pickup_type)) a$pickup_type = stops$pickup_type
    # if (!is.null(stops$drop_off_type)) a$drop_off_type = stops$drop_off_type
    # if (!is.null(stops$shape_dist_traveled)) a$shape_dist_traveled = stops$shape_dist_traveled
    # if (!is.null(stops$timepoint)) a$timepoint = stops$timepoint

    # tripids = unique(stops$trip_id)
    # pb <- txtProgressBar(0, length(tripids), 0, style = 3)
    # for (i in 1:length(tripids)) {
      # setTxtProgressBar(pb, i)
      # aa = lapply(a, function(x) x[stops$trip_id == tripids[i]])
      dbGetQuery(con,
        sprintf("INSERT INTO stop_times (%s) VALUES ('%s')",
                paste(names(a), collapse = ","),
                do.call(paste, c(a, list(sep = "','", collapse = "'),('")))))
    # }
    # close(pb)
  }
}

File: mybus.app/app/Console/Commands/UpdateGTFS.php
<?php

namespace App\Console\Commands;

use Illuminate\Console\Command;

use App;
use DB;
use HTTP_Request2;
use Zipper;

class UpdateGTFS extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'gtfs:update';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Update the GTFS static database';

    /**
     * Create a new command instance.
     *
     * @return void
     */
    public function __construct()
    {
        parent::__construct();
    }

    /**
     * Execute the console command.
     *
     * @return mixed
     */
    public function handle()
    {
        // Current versions:
        $versions = App\Version::orderBy('startdate', 'desc')->get();

        $req = new HTTP_Request2('https://api.at.govt.nz/v2/gtfs/versions');
        $url = $req->getUrl();
        $req->setHeader([
            'Ocp-Apim-Subscription-Key' => env('AT_API_KEY'),
        ]);

        $req->setMethod(HTTP_Request2::METHOD_GET);
        try {
            $response = $req->send();
            $new_versions = json_decode($response->getBody(), true)['response'];
        } catch (HttpException $ex) {
            echo $ex;
            dd();
        }


        foreach ($new_versions as $new_version) {

            $version = App\Version::firstOrNew($new_version);
            if (is_null($version->id)) {
                echo "*** Adding version $version->version\n\n";

                // New ID - add it and then insert new data!
                $version->save();

                // Execute download of GTFS files - if the folder doesn't exist!
                // TODO: use headers to check if gtfs.zip has been updated on AT's CDN
                if ( ! file_exists(storage_path() . '/gtfs.zip' )) {
                    echo ">>>>> Downloding GTFS data ... ";
                    file_put_contents(storage_path() . '/gtfs.zip',
                                      fopen("https://cdn01.at.govt.nz/data/gtfs.zip", "r"));
                    echo "done.\n";
                }


                // and then we need to do ALL the work!

                // -- AGENCIES
                echo "* Adding agencies ... ";
                $table = Zipper::make(storage_path() . '/gtfs.zip')->getFileContent('agency.txt');
                $table = array_map("str_getcsv", explode("\r\n", $table));
                if (count(end($table)) == 1) {
                    unset($table[count($table)-1]);
                }
                array_walk($table, function(&$a) use ($table) {
                    $a = array_combine($table[0], $a);
                });
                array_shift($table);

                foreach ($table as $a) {
                    $a = (object) $a;
                    $agency = App\Agency::where('agency_id', $a->agency_id)->firstOrCreate([
                        'agency_id' => $a->agency_id,
                        'name' => $a->agency_name,
                        'url' => $a->agency_url,
                        'timezone' => $a->agency_timezone,
                        'lang' => $a->agency_lang,
                        'phone' => $a->agency_phone
                    ]);
                }
                echo "done.\n";


                // -- ROUTES
                echo "* Adding routes ... ";
                $table = Zipper::make(storage_path() . '/gtfs.zip')->getFileContent('routes.txt');
                $table = array_map("str_getcsv", explode("\r\n", $table));
                if (count(end($table)) == 1) {
                    unset($table[count($table)-1]);
                }
                array_walk($table, function(&$a) use ($table) {
                    $a = array_combine($table[0], $a);
                });
                array_shift($table);

                foreach ($table as $r) {
                    $r = (object) $r;
                    if (str_contains($r->route_id, $version->version)) {
                        // Only go ahead if its the version we are interested in:
                        $route = App\Route::find($r->route_id);
                        if (is_null($route)) {
                            $route = new App\Route;
                            $route->create([
                                'id' => $r->route_id,
                                'route_id' => str_replace('-' . $version->version, '', $r->route_id),
                                'agency_id' => !empty($r->agency_id) ? $r->agency_id : null,
                                'short_name' => $r->route_short_name,
                                'long_name' => $r->route_long_name,
                                'desc' => !empty($r->route_desc) ? $r->route_desc : null,
                                'type_id' => $r->route_type,
                                'url' => !empty($r->route_url) ? $r->route_url : null,
                                'color' => !empty($r->route_color) ? $r->route_color : null,
                                'text_color' => !empty($r->route_text_color) ? $r->route_text_color : null,
                                'version_id' => $version->id
                            ]);
                        }
                    }
                }
                echo "done.\n";


                // -- TRIPS
                echo "* Adding trips ... ";
                $table = Zipper::make(storage_path() . '/gtfs.zip')->getFileContent('trips.txt');
                $table = array_map("str_getcsv", explode("\r\n", $table));
                if (count(end($table)) == 1) {
                    unset($table[count($table)-1]);
                }
                array_walk($table, function(&$a) use ($table) {
                    $a = array_combine($table[0], $a);
                });
                array_shift($table);

                foreach ($table as $t) {
                    $t = (object) $t;
                    if (str_contains($t->trip_id, $version->version)) {
                        $trip = App\Trip::find($t->trip_id);
                        if (is_null($trip)) {
                            $trip = new App\Trip;
                            $trip->create([
                                'id' => $t->trip_id,
                                'route_id' => $t->route_id,
                                'service_id' => $t->service_id,
                                'trip_id' => str_replace('-' . $version->version, '', $t->trip_id),
                                'headsign' => !empty($t->trip_headsign) ? $t->trip_headsign : null,
                                'short_name' => !empty($t->trip_short_name) ? $t->trip_short_name : null,
                                'direction_id' => !empty($t->direction_id) ? $t->direction_id : null,
                                'block_id' => !empty($t->block_id) ? $t->block_id : null,
                                'shape_id' => !empty($t->shape_id) ? $t->shape_id : null,
                                'wheelchair_accessible' =>
                                    !empty($t->wheelchair_accessible) ? $t->wheelchair_accessible : null,
                                'bikes_allowed' =>
                                    !empty($t->bikes_allowed) ? $t->bikes_allowed : null,
                                'version_id' => $version->id
                            ]);
                        }
                    }
                }
                echo "done.\n";


                // -- SHAPES
                echo "* Adding shapes ... ";
                // more difficult, because the file is large -- do it in chunks
                $shapes = fopen(storage_path() . '/shapes.csv', "w");
                fwrite($shapes, Zipper::make(storage_path() . '/gtfs.zip')->getFileContent('shapes.txt'));
                fclose($shapes);

                DB::statement('CREATE TABLE raw_shapes ( shape_id VARCHAR(255), ' .
                              'shape_pt_lat VARCHAR(255), shape_pt_lon VARCHAR(255), shape_pt_sequence INTEGER, ' .
                              'shape_dist_traveled DOUBLE PRECISION );');

                echo "-> table ";
                DB::statement("COPY raw_shapes FROM '" . storage_path() . "/shapes.csv' CSV HEADER;");
                unlink(storage_path() . '/shapes.csv');

                echo "-> shapes ... ";
                DB::statement(
                    "INSERT INTO shapes (id, shape_id, lat, lon, pt_sequence, dist_traveled, version_id) " .
                    "SELECT shape_id AS id, regexp_replace(shape_id, '-.+$', '') AS shape_id, " .
                    "shape_pt_lat AS lat, shape_pt_lon AS lon, shape_pt_sequence AS pt_sequence, " .
                    "shape_dist_traveled AS dist_traveled, $version->id AS version_id " .
                    "FROM raw_shapes WHERE shape_id LIKE '%$version->id'");

                DB::statement('DROP TABLE raw_shapes');
                echo "done.\n";


                // -- SHAPES
                echo "* Adding stops ... ";
                $table = Zipper::make(storage_path() . '/gtfs.zip')->getFileContent('stops.txt');
                $table = array_map("str_getcsv", explode("\r\n", $table));
                if (count(end($table)) == 1) {
                    unset($table[count($table)-1]);
                }
                array_walk($table, function(&$a) use ($table) {
                    $a = array_combine($table[0], $a);
                });
                array_shift($table);

                foreach ($table as $t) {
                    $t = (object) $t;
                    $stop = App\Stop::firstOrCreate([
                        'id' => $t->stop_id,
                        'code' => !empty($t->stop_code) ? $t->stop_code : null,
                        'name' => $t->stop_name,
                        'desc' => !empty($t->stop_desc) ? $t->stop_desc : null,
                        'lat' => $t->stop_lat,
                        'lon' => $t->stop_lon,
                        'zone_id' => !empty($t->zone_id) ? $t->zone_id : null,
                        'url' => !empty($t->url) ? $t->url : null,
                        'location_type' => !empty($t->location_type) ? $t->location_type : null,
                        'parent_station' => !empty($t->parent_station) ? $t->parent_station : null,
                        'timezone' => !empty($t->timezone) ? $t->timezone : null,
                        'wheelchair_boarding' =>
                            !empty($t->wheelchair_boarding) ? $t->wheelchair_boarding : null,
                    ]);
                }
                echo "done.\n";


                echo "Adding stop times ... csv ";
                // more difficult, because the file is large -- do it in chunks
                $stop_times = fopen(storage_path() . '/stop_times.csv', "w");
                fwrite($stop_times, Zipper::make(storage_path() . '/gtfs.zip')
                                      ->getFileContent('stop_times.txt'));
                fclose($stop_times);

                DB::statement('CREATE TABLE tmp ( trip_id VARCHAR(255), ' .
                              'arrival_time VARCHAR(255), departure_time VARCHAR(255), '.
                              'stop_id VARCHAR(255), stop_sequence INTEGER, stop_headsign VARCHAR(255), ' .
                              'pickup_type VARCHAR(255), drop_off_type VARCHAR(255), '.
                              'shape_dist_traveled DOUBLE PRECISION)');

                echo "-> table ";
                DB::statement("COPY tmp FROM '" . storage_path() . "/stop_times.csv' CSV HEADER;");
                unlink(storage_path() . '/stop_times.csv');

                echo "-> stop_times ... ";
                DB::statement(
                    "INSERT INTO stop_times (trip_id,arrival_time,departure_time,stop_id,stop_sequence,".
                                        "stop_headsign,pickup_type,drop_off_type,shape_dist_traveled) " .
                    "SELECT * FROM tmp WHERE trip_id LIKE '%-$version->version'");

                DB::statement('DROP TABLE tmp');
                echo "done.\n";


                // -- CALENDAR
                echo "* Adding calendars ... ";
                $table = Zipper::make(storage_path() . '/gtfs.zip')->getFileContent('calendar.txt');
                $table = array_map("str_getcsv", explode("\r\n", $table));
                if (count(end($table)) == 1) {
                    unset($table[count($table)-1]);
                }
                array_walk($table, function(&$a) use ($table) {
                    $a = array_combine($table[0], $a);
                });
                array_shift($table);

                foreach ($table as $c) {
                    $c = (object) $c;
                    if (str_contains($c->service_id, $version->version)) {
                        $calendar = App\Calendar::find($c->service_id);
                        if (is_null($calendar)) {
                            $calendar = new App\Calendar;
                            $calendar->create([
                                'id' => $c->service_id,
                                'service_id' => str_replace('-' . $version->version, '', $c->service_id),
                                'monday' => (bool)$c->monday,
                                'tuesday' => (bool)$c->tuesday,
                                'wednesday' => (bool)$c->wednesday,
                                'thursday' => (bool)$c->thursday,
                                'friday' => (bool)$c->friday,
                                'saturday' => (bool)$c->saturday,
                                'sunday' => (bool)$c->sunday,
                                'start_date' => $c->start_date,
                                'end_date' => $c->end_date,
                                'version_id' => $version->id
                            ]);
                        }
                    }
                }
                echo "done.\n";


                // -- CALENDAR DATES
                echo "* Adding calendar dates ... ";
                $table = Zipper::make(storage_path() . '/gtfs.zip')->getFileContent('calendar_dates.txt');
                $table = array_map("str_getcsv", explode("\r\n", $table));
                if (count(end($table)) == 1) {
                    unset($table[count($table)-1]);
                }
                array_walk($table, function(&$a) use ($table) {
                    $a = array_combine($table[0], $a);
                });
                array_shift($table);

                foreach ($table as $cd) {
                    $cd = (object) $cd;
                    if (str_contains($cd->service_id, $version->version)) {
                        $calendar_date = App\CalendarDate::firstOrCreate([
                            'service_id' => $cd->service_id,
                            'date' => $cd->date,
                            'exception_type' => $cd->exception_type
                        ]);
                    }
                }
                echo "done.\n";
            }
        }
    }
}
